src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubRDDIterator.java

```diff
@@ -0,0 +1,169 @@
+                           boolean autoAcknowledge, Credentials credentials) {
+    this.config = config;
+    this.context = context;
+    this.batchDuration = batchDuration;
+    this.credentials = credentials;
+    this.startTime = batchTime.milliseconds();
+    this.autoAcknowledge = autoAcknowledge;
+    subscriptionFormatted = ProjectSubscriptionName.format(this.config.getProject(), this.config.getSubscription());
+    receivedMessages = new ConcurrentLinkedDeque<>();
+  }
+
+  @Override
+  public boolean hasNext() {
+    // Complete processing of acknowledged messages before finishing the batch.
+    if (!receivedMessages.isEmpty()) {
+      return true;
+    }
+
+    long currentTimeMillis = System.currentTimeMillis();
+    if (currentTimeMillis >= (startTime + batchDuration)) {
+      LOG.debug("Time exceeded for batch. Total time is {} millis. Total messages returned is {} .",
+                currentTimeMillis - startTime, messageCount);
+      return false;
+    }
+
+    try {
+      List<ReceivedMessage> messages = fetchAndAck();
+      //If there are no messages to process, continue.
+      if (messages.isEmpty()) {
+        LOG.debug("No more messages. Total messages returned is {} .", messageCount);
+        return false;
+      }
+      receivedMessages.addAll(messages);
+      return true;
+    } catch (IOException e) {
+      throw new RuntimeException("Error reading messages from Pub/Sub. ", e);
+    }
+  }
+
+  @Override
+  public PubSubMessage next() {
+    if (receivedMessages.isEmpty()) {
+      // This should not happen, if hasNext() returns true, then a message should be available in queue.
+      throw new IllegalStateException("Unexpected state. No messages available.");
+    }
+
+    ReceivedMessage currentMessage = receivedMessages.poll();
+    messageCount += 1;
+    return new PubSubMessage(currentMessage);
+  }
+
+  private SubscriberStub buildSubscriberClient() throws IOException {
+    SubscriberStubSettings.Builder builder = SubscriberStubSettings.newBuilder();
+    if (credentials != null) {
+      builder.setCredentialsProvider(FixedCredentialsProvider.create(credentials));
+    }
+    builder.getSubscriptionSettings().setRetrySettings(PubSubSubscriberUtil.getRetrySettings());
+    return GrpcSubscriberStub.create(builder.build());
+  }
+
+  private List<ReceivedMessage> fetchAndAck() throws IOException {
+    if (this.subscriber == null) {
+      this.subscriber = buildSubscriberClient();
+      context.addTaskCompletionListener(context1 -> {
+        if (subscriber != null && !subscriber.isShutdown()) {
+          subscriber.shutdown();
+          try {
+            subscriber.awaitTermination(30, TimeUnit.SECONDS);
+          } catch (InterruptedException e) {
+            LOG.warn("Exception in shutting down subscriber. ", e);
+          }
+        }
+      });
+      this.pullRequest = PullRequest.newBuilder()
+        .setMaxMessages(MAX_MESSAGES)
```

Comment:
this seems like it should be configurable, especially since we don't know the message size. Though looks like we don't have it right now, so can be done in a future PR.

Reply:
Makes sense, Kafka has a similar option. Created https://cdap.atlassian.net/browse/PLUGIN-1598 to track this.